observers.md

```diff
@@ -0,0 +1,87 @@
+
+In the example below, we're using a callback to initiate an action after creating a new user. We're mixing `MessageMailer` into the `User` class:
+
+```
+class User
+  after_create :notify
+
+  protected
+    def notify
+      MessageMailer.send_welcome_email(self).deliver
+    end
+end
+```
+
+As a better alternative, by using observers in the example below, we extract the `MessageMailer` logic out of `User` and put it in its own class, `UserObserver`, that watches for an event and dispatches an action accordingly:
+
+```
+class UserObserver < ActiveRecord::Observer
+  def after_create(user)
+    MessageMailer.send_welcome_email(user)
+  end
+end
+```
+
+Observers can also watch multiple classes:
+
+```
+class AuditObserver < ActiveRecord::Observer
+	observe :account, :balance
+
+	def after_update(record)
+		AuditTrail.new(record, "UPDATED")
+	end
+end
+```
+
+Obervers are especially helpful in cases like the above where there are multiple models comingled and we want to hook into all of them to watch for some event. We can pull logic out of models where it doesn't belong, put it in its own separate place, observe actions, and still affect change as needed.
+
+### Accounting for unseen side effects
+
+One thing to consider with observers is that we now have events that are not readily apparent in our observed class. Due to these unseen side effects, it can be less clear what's happening, and if there's a bug, it can be hard to track, since there is no direct link to the thing that is observing it.
+
+To mitigate this potential downside, we can write observers that reference the model they are observing. One way to do that is with `logger`. In this example from the [Rails documentation](http://api.rubyonrails.org/v2.3/classes/ActiveRecord/Observer.html), the observer uses `ActiveSupport::Logger` to log when specific callbacks are triggered:
+
+```
+class ContactObserver < ActiveRecord::Observer
+    def after_create(contact)
+      contact.logger.info('New contact added!')
+    end
+
+    def after_destroy(contact)
+      contact.logger.warn("Contact with an id of #{contact.id} was destroyed!")
+    end
+  end
+```
+
+### Staying in sync
+
+Another thing to consider is that since we're dealing with application state, it's possible that as the application transitions from one state to another, our observers can become out of sync.
+
+For example, payment processing transactions usually have many steps. Consider a case where a transaction is created but the payment is then rejected. If we have an observer with an `after_create` callback, that callback is going to fire after `create`, regardless of whether that transaction is eventually rolled back.
```

Comment:
Trying to find where I picked up this concept. I think what I'm trying to say here is that if there are multiple actions inside an `after_create` and one of them fails, the others will continue on... in any case, you're right that this is a more in-depth topic that requires more research and its own post.